src/common/Throttle.h

```diff
@@ -330,4 +332,71 @@ class OrderedThrottle {
   uint32_t waiters = 0;
 };
 
+
+class Bucket {
+  CephContext *cct;
+  const std::string name;
+  std::atomic<int64_t> remain = { 0 }, max = { 0 };
+  Mutex lock;
+  list<Cond*> cond;
+
+public:
+  Bucket(CephContext *cct, const std::string& n, int64_t m = 0);
+  ~Bucket();
+
+private:
+  bool _should_wait(int64_t c) const {
+    return (remain <= 0);
+  }
+
+  bool _wait(int64_t c);
+
+public:
+  int64_t get_max() const { 
+    return max; 
+  }
+  int64_t get_current() const {
+    return remain;
+  }
+  void set_max(int64_t m);
+  bool get(int64_t c = 1);
+  int64_t put(int64_t c = 1);
+};
+
+
+class TokenBucketThrottle {
+  CephContext *m_cct;
+  Bucket throttle;
+  int64_t m_avg;
+  SafeTimer *m_timer;
+  Mutex *m_timer_lock;
+  FunctionContext *m_token_ctx;
+
+public:
+  TokenBucketThrottle(CephContext *cct,
+		      int64_t capacity,
+		      int64_t avg,
+		      SafeTimer *timer,
+		      Mutex *timer_lock)
+  : m_cct(cct), throttle(m_cct, "token_bucket_throttle", capacity),
+    m_avg(avg), m_timer(timer), m_timer_lock(timer_lock)
+  {
+      Mutex::Locker timer_locker(*m_timer_lock);
+      add_tokens();
+  }
+  ~TokenBucketThrottle(){
+    Mutex::Locker timer_locker(*m_timer_lock);
+    cancel_tokens();
+  }
+
+private:
+  void add_tokens();
+  void cancel_tokens();
+
+public:
+  int get(int64_t c);
```

Comment:
The IO path needs to be non-blocking, so this would need to support ```Context``` callback. Since it would be wasteful to construct a ```Context``` for the allowed case, you could provide a templated ```get``` method that had a parameters for the callback class and function. If a throttle is required, it could create a ```Context``` to invoke the provided class/function when tokens are available and return false; otherwise it returns true to indicate that throttling isn't required.